File: databasyrepo/databasyrepo/models/core/nodes.py
from databasyrepo.models.core.serializing import Serializable

__author__ = 'Marboni'

class Node(Serializable):
    def __init__(self, id=None):
        super(Node, self).__init__()
        self.set('_id', id)

    def fields(self):
        return {
            '_id': basestring,
            }

    @property
    def id(self):
        return self.val('_id')

    def insert_item(self, field, index, item):
        lst = self._iter_val(field)
        lst.insert(index, item)

    def append_item(self, field, item):
        lst = self._iter_val(field)
        lst.insert(self.items_count(field), item)

    def delete_item(self, field, index):
        lst = self._iter_val(field)
        return lst.pop(index)

    def item_index(self, field, item):
        lst = self._iter_val(field)
        return lst.index(item)

    def items_count(self, field):
        lst = self._iter_val(field)
        return len(lst)

    def _iter_val(self, field):
        value = self.val(field)
        try:
            iter(value)
        except TypeError:
            raise ValueError('Field %s is not iterable.' % field)
        return value

    def ref(self):
        #noinspection PyUnresolvedReferences
        return NodeRef(ref_id=self.id, ref_code=self.code())

class NodeRef(Serializable):
    def fields(self):
        return {
            'ref_id': basestring,
            'ref_code': basestring
        }

    @property
    def ref_id(self):
        return self['ref_id']

    def ref_node(self, model):
        node = model.node(self.ref_id)
        node_code = node.code()
        if node_code != self['ref_code']:
            raise ValueError('Node type (%s) not equal to the type in node reference (%s).' % (node_code, self['ref_code']))
        return node

File: databasyrepo/databasyrepo/static/src/databasy/model/core/nodes.js
databasy.model.core.nodes.Node = databasy.model.core.serializing.Serializable.extend({
    init:function(params) {
        this._super(params);
    },
    fields:function () {
        return this._super().concat(
            '_id'
        )
    },
    id:function () {
        return this.val('_id');
    },
    insert_item:function (field, index, item) {
        var lst = this._iter_val(field);
        lst.splice(index, 0, item);
    },
    append_item:function(field, item) {
        var lst = this._iter_val(field);
        lst.splice(this.items_count(field), 0, item);
    },
    delete_item: function(field, index) {
        var lst = this._iter_val(field);
        var item_to_delete = lst[index];
        lst.splice(index, 1);
        return item_to_delete;
    },
    item_index: function(field, item) {
        var lst = this._iter_val(field);
        var index = $.inArray(item, lst);
        if (index == -1) {
            throw new Error('Item not found in the list "' + field + '".')
        } else {
            return lst[index];
        }
    },
    items_count:function(field) {
        var lst = this._iter_val(field);
        return lst.length;
    },
    _iter_val:function (field) {
        var value = this.val(field);
        if (Object.prototype.toString.call(value) !== '[object Array]') {
            throw new Error('Field "' + field + '" is not iterable.');
        }
        return value;
    },
    ref:function() {
        return new databasy.model.core.nodes.NodeRef({ref_id:this.id(), ref_code:this.code()})
    }
});


databasy.model.core.nodes.NodeRef = databasy.model.core.serializing.Serializable.extend({
    fields:function () {
        return this._super().concat(
            'ref_id',
            'ref_code'
        )
    },
    ref_node:function (model) {
        return model.node(this.val('ref_id'));
    }
}, {
    CODE:'core.nodes.NodeRef'
});
